### ISSUE TITLE
index_put modifies incorrect dimension when boolean mask is followed by null index

### ISSUE THREAD

**Issue by manuelcandales** [pytorch/pytorch#158413]
### 🐛 Describe the bug

This needs to be explained with an example:
```python
>>> import torch
>>> x = torch.tensor([[1, 2], [3, 4]])
>>> m = torch.tensor([True, False])
>>> x
tensor([[1, 2],
        [3, 4]])
>>> torch.ops.aten.index(x, [m])
tensor([[1, 2]])  # correct
>>> torch.ops.aten.index(x, [m, None])
tensor([[1, 2]])  # correct
>>> torch.ops.aten.index_put(x, [m], torch.tensor(0))
tensor([[0, 0],   # correct
        [3, 4]])
>>> torch.ops.aten.index_put(x, [m, None], torch.tensor(0))
tensor([[0, 2],   # incorrect, it should be [[0, 0], [3, 4]] like the last command
        [0, 4]])
```

Notice that we don't have this problem if instead of using a boolean mask, we use an integer tensor.
```python
>>> ix = torch.tensor([0,])  # equivalent to boolean mask m when applied on dimension of size 2
>>> torch.ops.aten.index_put(x, [ix], torch.tensor(0))
tensor([[0, 0],
        [3, 4]])
>>> torch.ops.aten.index_put(x, [ix, None], torch.tensor(0))
tensor([[0, 0],
        [3, 4]])
```

### Versions

nightly

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim

**Comment by ngimel** [comment]
This is due to incorrect masked_fill dispatch https://github.com/pytorch/pytorch/blob/beb52f5c0ac8efddeafa862f274bc247db989695/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L978, when the src tensor is more than 1 elem everything is correct.
